org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.java

```diff
@@ -1019,6 +1023,25 @@ protected IResourceServiceProvider getResourceServiceProvider(URI uri) {
 		options.setCancelIndicator(cancelIndicator);
 		return renameService.prepareRename(options);
 	}
+	
+	/**
+	 * @since 2.25
+	 */
+	public CompletableFuture<List<FoldingRange>> foldingRange(FoldingRangeRequestParams params) {
+		return requestManager.runRead(cancelIndicator -> foldingRange(params, cancelIndicator));
+	}
+	
+	/**
+	 * @since 2.25
+	 */
+	protected List<FoldingRange> foldingRange(FoldingRangeRequestParams params, CancelIndicator cancelIndicator) {
+		URI uri = getURI(params.getTextDocument());
+		FoldingRangeService foldingRangeService = getService(uri, FoldingRangeService.class);
+		if (foldingRangeService == null) {
+			throw new UnsupportedOperationException();
```

Comment:
I think an empty list would be more user friendly if the service is not implemented